File: MyVideoGameList.Server/Security/SecurityHeaders.cs
namespace MyVideoGameList.Server.Security;

/// <summary>
/// Response headers that state what this API's answers may be used for.
/// </summary>
/// <remarks>
/// <para>
/// Everything here serves JSON, so the policy is the strictest one there is: a document that
/// loads nothing and may not be framed. That covers the two ways a JSON endpoint gets turned
/// into something else — a browser sniffing a response into a script, and a page framing an
/// answer meant to be read by code.
/// </para>
/// <para>
/// The HTML has its own set, because it is served by the other process
/// (<c>docs/decisions/0003-two-process-deployment.md</c>) and a document that loads nothing would
/// be a blank page. See <c>myvideogamelist.client/src/lib/securityHeaders.ts</c>.
/// </para>
/// </remarks>
public sealed class SecurityHeadersMiddleware(RequestDelegate next)
{
    /// <summary>
    /// A JSON API is not a document: it loads nothing, embeds nothing and frames nothing.
    /// </summary>
    private const string ContentSecurityPolicy = "default-src 'none'; frame-ancestors 'none'";

    /// <summary>
    /// Registers the headers to be written when the response starts, rather than writing them
    /// now. The difference matters on the error path: the exception handler clears the response
    /// before writing its own, headers included, so anything set on the way in is gone by the
    /// time a 500 or a 502 goes out - which are the answers least worth sending bare.
    /// </summary>
    public async Task InvokeAsync(HttpContext context)
    {
        context.Response.OnStarting(static state =>
        {
            Apply((HttpContext)state);
            return Task.CompletedTask;
        }, context);

        await next(context);
    }

    internal static void Apply(HttpContext context)
    {
        var headers = context.Response.Headers;

        // Without this a response whose body looks like script can be loaded as one, whatever
        // its content type says.
        headers.XContentTypeOptions = "nosniff";

        // An API request is never a navigation, so there is no referrer worth sending. Paths
        // here carry usernames and game ids.
        headers["Referrer-Policy"] = "no-referrer";

        // The header and its modern replacement, because the two are read by different browsers
        // and neither is universal on its own.
        headers.XFrameOptions = "DENY";

        if (!IsApiDocumentation(context.Request.Path))
            headers.ContentSecurityPolicy = ContentSecurityPolicy;
    }

    /// <summary>
    /// The OpenAPI document and the Scalar UI in front of it, which are mapped in Development
    /// only. They are the one thing this process serves that <em>is</em> a document, and the
    /// policy above would leave it blank.
    /// </summary>
    private static bool IsApiDocumentation(PathString path) =>
        path.StartsWithSegments("/scalar") || path.StartsWithSegments("/openapi");
}

public static class SecurityHeaders
{
    public static IApplicationBuilder UseApiSecurityHeaders(this IApplicationBuilder app) =>
        app.UseMiddleware<SecurityHeadersMiddleware>();
}

File: myvideogamelist.client/src/lib/securityHeaders.ts
/**
 * Headers every server-rendered document carries, whatever it renders.
 *
 * The API states its own set (`MyVideoGameList.Server/Security/SecurityHeaders.cs`), and the two
 * are deliberately different: JSON may load nothing, whereas a page that loaded nothing would be
 * blank. They are two processes serving one site — see ADR 0003 — so neither set covers the other,
 * and a header added to one is worth a look at the other.
 *
 * **What is not here.** A content policy naming `script-src` or `style-src` is the valuable half of
 * CSP and it is missing, because React Router hydrates from an inline script: restricting scripts
 * means a per-request nonce handed to `<Scripts nonce>`, and a wrong one is a blank page rather
 * than a warning. `frame-ancestors` is the part that needs no nonce, so it ships now and the rest
 * is a change of its own.
 */
export const SECURITY_HEADERS: Readonly<Record<string, string>> = {
    /** A response whose body looks like script must not be loadable as one. */
    'X-Content-Type-Options': 'nosniff',

    /**
     * The origin, never the path, once the destination is somebody else. Paths here carry
     * usernames and the games a person tracks, and the outbound links are to stores and IGDB.
     */
    'Referrer-Policy': 'strict-origin-when-cross-origin',

    /**
     * Framing is how a signed-in page's controls get clicked by a page the reader thinks they are
     * on instead. Both spellings, because they are read by different browsers.
     */
    'X-Frame-Options': 'DENY',
    'Content-Security-Policy': "frame-ancestors 'none'",
};

/**
 * Applies them to a response's headers, in place.
 *
 * Called once for every document response, including the ones a thrown `Response` produces — which
 * is the reason this happens in `entry.server.tsx` rather than in each route's `headers` export.
 * A throw bypasses that export, so a per-route policy would leave every 404 uncovered.
 */
export function applySecurityHeaders(headers: Headers): Headers {
    for (const [name, value] of Object.entries(SECURITY_HEADERS)) headers.set(name, value);
    return headers;
}
